lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2015-2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package opsgenie
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/gravitational/teleport-plugins/lib"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+const (
+	ogMaxConns    = 100
+	ogHTTPTimeout = 10 * time.Second
+
+	ogAlertKeyPrefix = "teleport-access-request"
+)
+
+var alertBodyTemplate = template.Must(template.New("alert body").Parse(
+	`{{.User}} requested permissions for roles {{range $index, $element := .Roles}}{{if $index}}, {{end}}{{ . }}{{end}} on Teleport at {{.Created.Format .TimeFormat}}.
+{{if .RequestReason}}Reason: {{.RequestReason}}{{end}}
+{{if .RequestLink}}To approve or deny the request, proceed to {{.RequestLink}}{{end}}
+`,
+))
+var reviewNoteTemplate = template.Must(template.New("review note").Parse(
+	`{{.Author}} reviewed the request at {{.Created.Format .TimeFormat}}.
+Resolution: {{.ProposedState}}.
+{{if .Reason}}Reason: {{.Reason}}.{{end}}`,
+))
+var resolutionNoteTemplate = template.Must(template.New("resolution note").Parse(
+	`Access request has been {{.Resolution}}
+{{if .ResolveReason}}Reason: {{.ResolveReason}}{{end}}`,
+))
+
+// OpsgenieClient is a wrapper around resty.Client.
+type OpsgenieClient struct {
+	client *resty.Client
+	// webProxyURL is the address used when building the bodies of the alerts
+	// allowing links to the access requests to be built
+	webProxyURL      *url.URL
+	clusterName      string
+	defaultSchedules []string
+	Priority         String
+}
+
+type OpsgenieClientConfig struct {
+	APIKey           string
+	APIEndpoint      string
+	defaultSchedules []string
+	Priority         String
+}
+
+func NewOpsgenieClient(conf OpsgenieClientConfig, clusterName, webProxyAddr string) (OpsgenieClient, error) {
```

Comment:
I would make `clusterName` and `webProxyAddr` a part of the config. Also, return a pointer.
```suggestion
func NewOpsgenieClient(conf OpsgenieClientConfig, clusterName, webProxyAddr string) (*OpsgenieClient, error) {
```